src/main/java/com/pusher/client/connection/websocket/WebSocketConnection.java

```diff
@@ -257,6 +264,39 @@ public void onClose(final int code, final String reason, final boolean remote) {
             return;
         }
 
+        //Reconnection logic
+        if(state == ConnectionState.CONNECTED || state == ConnectionState.CONNECTING){
+
+            if(reconnectAttempts < MAX_CONNECTION_ATTEMPTS){
+                tryReconnecting();
+            }
+            else{
+                updateState(ConnectionState.DISCONNECTING);
+                cancelTimeoutsAndTransitonToDisconnected();
+            }
+            return;
+        }
+
+        if (state == ConnectionState.DISCONNECTING){
+            cancelTimeoutsAndTransitonToDisconnected();
+        }
+    }
+
+    private void tryReconnecting() {
+        reconnectAttempts++;
+        updateState(ConnectionState.RECONNECTING);
```

Comment:
I followed the Swift lib for this flow where Reconnecting is just a step before Connecting. Also makes sense because Reconnecting will do progressively longer waits